### ISSUE TITLE
Library

### ISSUE THREAD

**Issue by esteinig** [esteinig/beastling#1]
Ideas:

* production grade phylodynamics library in Python
* bacterial phylodynamics module:
    * clonal expansion coalescent model (`CaveDive`)
    * birth-death skyline models (`BDSky` and `MTBD`)
* `SARS-CoV-2` models for variants and national / subnational configs
* integration with Nextflow to run BEAST + BEAGLE on GPU

**Comment by esteinig** [comment]
I'd like to make model classes that can be inherited to define specific sub-models. Models represent configurations of the BEAST XML and are built from XML blocks (`beastling.blocks`)

```python
from beastling.models import BirthDeathSkylineSerial

class MRSA(BirthDeathSkylineSerial):
    pass

```

Each model inherits from the base `Beastling` class.

**Comment by esteinig** [comment]
For the terminal client configuration, it'd be cool if we could use `YAML` templates that configure the model (and can be default configured by users) and let users modify arguments not provided in the file for on-the-fly command line configuration.

I have a template with the base `click` library for this, but would rather use `Typer` so need to check out how to modify the `context` object (since it depends on `click` it should be managable)

**Comment by esteinig** [comment]
Something like this in a very basic sense ^^
![schema](https://user-images.githubusercontent.com/12873366/124339121-69957000-dbef-11eb-917f-0598d3d5f407.png)

**Comment by Wytamma** [comment]
Very cool 😎 I like this idea. I think having the model specification/XML generation separated from any deployment would be very useful for 3rd parties wanting to use this library. Also would it be possible to cut out the YAML and have users directly specify the models in python (eg dataclass)? This would enable easy dynamic model creation.

**Comment by esteinig** [comment]
Yeah awesome, we should do that! I think the `YAML` thing still comes from letting users specify their own template for the command line interface, e.g. when users provide a template and configure an empty string for a parameter, the CLI will prompt the user for that specific param (in interactive mode) or it is required to be specified on the command line.

Does that makes sense? That way users can specify their CLI commands for specific models in a template file.

```yaml
param1: default value
param2: ''
param3: null
```

In the CLI:

Interactive, prompts user for `param2`
```
beastling -i -t template.yaml
```

Required command line:

```
beastling --param2 test
```

Can you explain the dynamic config? I'm not too familiar! Maybe I'm thinking too much on the user end right now.

**Comment by Wytamma** [comment]
Yes! I think that’s a great way to do it. I’m just thinking more about it as a library that is consumed by other apps than an application used by users, but both are important. Programmatic control of BEAST config would be very useful and would open up an interface between python and BEAST.
